cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -306,75 +327,201 @@ public String toDisplayString(Subject subject) {
+                throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN,
+                        "Cannot validate token using user info endpoint", e);
+            }
+        }
+
+        s.getPrincipals().remove(raw);
+        for (Principal p : validatedPrincipals) {
+            s.getPrincipals().add(p);
+        }
+
+        // TODO - oidcScope not assigned yet
+        AuthorizationToken authToken = new AuthorizationToken(challengeType, credentials, oidcDomains, oidcScope);
+        s.getPublicCredentials().add(authToken);
+    }
+
+    private List<Principal> validateWithPubKey(URI jwtIssuer, String challengeType,String credentials)
+            throws MalformedURLException, InvalidJwtException, MalformedClaimException {
+        VerificationKeyResolver httpsJwksKeyResolver = getHttpsJwksVerificationKeyResolver(jwtIssuer, challengeType);
+        JwtConsumer jwtConsumer = new JwtConsumerBuilder()
+                .setRequireExpirationTime()
+                .setExpectedIssuers(true, jwtIssuer.toString())
+                .setVerificationKeyResolver(httpsJwksKeyResolver)
+                .setJwsAlgorithmConstraints(new AlgorithmConstraints(AlgorithmConstraints.ConstraintType.PERMIT, // whitelist
+                        AlgorithmIdentifiers.RSA_USING_SHA256)) // RS256 only
+                .build(); // create the JwtConsumer instance;
+
+        //  Validate the JWT and process it to the Claims
+        JwtClaims jwtClaims = jwtConsumer.processToClaims(credentials);
+        log.debug("JWT validation succeeded! " + jwtClaims);
+
+        String sub = jwtClaims.getClaimValue("sub", String.class);
+
+        List<Principal> result = new ArrayList<>();
+        OpenIdPrincipal oip = new OpenIdPrincipal(jwtIssuer.toURL(), sub);
+        result.add(oip);
+
+        if (jwtClaims.getClaimValueAsString("preferred_username") != null) {
+            result.add(new HttpPrincipal(jwtClaims.getClaimValueAsString("preferred_username")));
+        }
+        log.debug("Validated user via issuer pub key: " + oip);
+        return result;
+    }
+
+    private static List<Principal> validateWithUserInfo(AuthorizationTokenPrincipal raw, URL issuerURL)
+            throws ResourceAlreadyExistsException, ResourceNotFoundException, IOException, InterruptedException {
+        HttpGet get = new HttpGet(issuerURL, true);
+        get.setRequestProperty("authorization", raw.getHeaderValue());
+        get.prepare();
+
+        InputStream istream = get.getInputStream();
+        String str = StringUtil.readFromInputStream(istream, "UTF-8");
+        JSONObject json = new JSONObject(str);
+        String sub = json.getString("sub");
+        String username = null;
+        if (json.has("preferred_username")) {
+            // jwt token
+            username = json.getString("preferred_username");
+        } else {
```

Comment:
`else if {` ??

and since we ran into the before (well, Rubin did), maybe a final `else {` with a debug statement that there is no username would be a good addition here.

also note that on line 432 you only check for `(jwtClaims.getClaimValueAsString("preferred_username)`... is that more strictly defined?

Reply:
Left the `name` option strictly for the CADC case.